### PR TITLE
Allow archiving of questions.

### PR DESCRIPTION
**Pull request by ndmckinley**
### Description
Introduces three new actions for questions - archive, restore, and discard draft.

### Checklist
- [ ] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### Issue(s)
Implements #997.

### CODE REVIEW

**Review by bion — APPROVED** [review]

**Review by bion — CHANGES_REQUESTED** [review]
accidentally approved the wrong PR, sorry. still reviewing this

File: universal-application-tool-0.0.1/app/models/Version.java [hunk 1]
```diff
@@ -76,4 +82,36 @@ public Instant getSubmitTime() {
         .filter(q -> q.getQuestionDefinition().getName().equals(name))
         .findAny();
   }
+
+  public ImmutableList<String> getTombstonedProgramNames() {
```
Comment (bion) [inline universal-application-tool-0.0.1/app/models/Version.java:86]: Suggestion (blocking): please add a comment explaining what it means for a question or program to have its name in the tombstoned list.

File: universal-application-tool-0.0.1/app/models/Version.java [hunk 2]
```diff
@@ -76,4 +82,36 @@ public Instant getSubmitTime() {
...
+      return ImmutableList.of();
+    }
+    return ImmutableList.copyOf(this.tombstonedQuestionNames);
+  }
+
+  public boolean addTombstoneForQuestion(Question question) {
+    if (this.tombstonedQuestionNames == null) {
+      this.tombstonedQuestionNames = new ArrayList<>();
+    }
+    if (this.tombstonedQuestionNames.contains(question.getQuestionDefinition().getName())) {
```
Comment (bion) [inline universal-application-tool-0.0.1/app/models/Version.java:104]: Suggestion (blocking): calling `contains` with the question and program name in here and outside this class looks like an implementation detail leaking out of its abstraction. Please implement methods `public boolean questionIsTombstoned(QuestionDefinition qd)` and `public boolean programIsTombstoned(ProgramDefinition pd)` that encapsulate that check.

**Review by bion — COMMENTED** [review]
Question (blocking): "archive" and "delete" appear to be used interchangeably in this PR. Does deletion actually take place?

**Comment by ndmckinley** [comment]
Sure thing!  I actually hadn't meant to send this out for review yet - it needs comments and tests still - but I'll add your suggestions to the commit also.

File: universal-application-tool-0.0.1/app/models/Version.java (hunk 1 above)
Comment (ndmckinley) [reply universal-application-tool-0.0.1/app/models/Version.java:96]: Done.

File: universal-application-tool-0.0.1/app/models/Version.java (hunk 2 above)
Comment (ndmckinley) [reply universal-application-tool-0.0.1/app/models/Version.java]: Done.

**Comment by ndmckinley** [comment]
> Question (blocking): "archive" and "delete" appear to be used interchangeably in this PR. Does deletion actually take place?

Sort of - it's "deleted" from future versions but still accessible to program admins.  I think they're interchangeable.  You have a preference for one or the other?

**Comment by bion** [comment]
> Sure thing! I actually hadn't meant to send this out for review yet - it needs comments and tests still - but I'll add your suggestions to the commit also.

oh! sorry about that! i'm not sure how i got turned around

**Review by bion — APPROVED** [review]

**Comment by bion** [comment]
> You have a preference for one or the other?

I like "archived", imo it suggests that it still exists in the system somewhere though is not currently in use

**Pull request merged**